packages/xstate-solid/src/types.ts

```diff
@@ -0,0 +1,72 @@
+import type {
+  AnyStateMachine,
+  AreAllImplementationsAssumedToBeProvided,
+  EventObject,
+  InternalMachineOptions,
+  InterpreterOptions,
+  State,
+  StateConfig,
+  StateSchema,
+  TypegenDisabled,
+  Typestate
+} from 'xstate';
+
+type StateObject<
+  C,
+  E extends EventObject,
+  S extends StateSchema<C> = any,
+  T extends Typestate<C> = { value: any; context: C },
+  R = TypegenDisabled
+> = State<C, E, S, T, R>;
+
+export type CheckSnapshot<Snapshot> = Snapshot extends State<
+  infer C,
+  infer E,
+  infer S,
+  infer T,
+  infer R
+>
+  ? StateObject<C, E, S, T, R>
```

Comment:
Isn't this equivalent to just this?
```suggestion
  ? State<C, E, S, T, R>
```

Reply:
Is this not what you wanted with a `POJOState`? I may have misunderstood, but what I have now creates an alias that signifies that `State` in an object and not a class. The shape of both are the same during runtime

Reply:
This doesn't change the underlying type of the `State` though, it's still treated as an instance of a class:
[TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgQQHYE8DKMCGMApgDQpwCiAbkRjAPIBGAVkQMYxmp6Gnk-EcbABZEQ+LnAAqWMEQDmtACLAAzvkYAbIgBNJMuaoLEyAXzgAzKBBBwARAA8jvOwG44Aeg9wYsoqoAuOAAWADoAZnDQgDYkJF85OAEiJlYOAB5yAGFJCjgiB2IMHVVKGjpU9k5+fMLaEqTjIiFRcXSsgD44AF44fGx9WqKGg38m9q7ehDgqfE0AVyIg-qx3Ngg6ApggrLhTSQAlHuk-RQwVdS1dJEnG3naSSkecR6lHg47XOILIWB8-OBZURsADWOAw+DAqmEEBg6XBkOhsNuCKhMPgW3qpWSmVQwAwFiIUEBknxhOJFFJBKJSSp5OkdJpBxu5AA-HdiJUMjknrSGXAPuQgqikTAvkgdOxNPgoEQ4OsMEY4JBmBBkgBGHbAsEQtGw+HjabreZ0IIYeYgRg00wdT5xBVK-BsNhEMDwXoAChAuCaQUwPt4cAAPnBzZaiQBKHpdZB4ixwL0B4hwfHODAuiDx5JR2OoODe5JfPOebwAPVZ5HMRE0qjludQXjgEAgUMeMGEajgkos+OAxE0WDgAHcNgByeBWnxQRZwUSy8ioAtNIt5xvlytIUzixsAARgqgAtD8tkeoNYoEgnS63ckPSq1U11RGvkA)

Reply:
Even if I change StateObject to be a unique type without an alias, it still doesn't produce an error. [TS playground ](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgQQHYE8DKMCGMApgDQpwCiAbkRjAPIBGAVkQMYxmp6Gnk-EcbABZEQ+LnAAqWMEQDmtACLAAzvkYAbIgBNJMuaoLEyAXzgAzKBBBwARAA8jvOwG44Aeg9wYsoqoAuOAAWADoAZnDQgDYkJF85OAEiJlYOAB5yAGFJCjgiB2IMHVVKGjpU9k5+fMLaEqTjIiFRcXSsgD44AF44fGx9WqKGg38m9q7ehDgqfE0AVyIg-qx3Ngg6ApggrLhTSQAlHuk-RQwVdS1dJEnEcnXNwp2kUziCyFgfPzgs0TYAaxwGHwYFUwggMHSQJBYIht2hoPB8C29VKyUyqGAGAsRCgP0kWJxeIoBOxuKSpKJ0kp5ION3IAH5GrxKhkcpQSBTqXADh1yEEEbCYK44kgdOxNPgoEQ4A8jHBIMwIMkAIw7P6A4GIiFQ8bTdbzOhBDDzECMcmmDodEVIOXwfBsNhEMAwZLHAAUIFwTSCmG9vDgAB84CazbiAJQ9LrITEWOCe-3EOBY5wYJ0QOPJSMx1BwL3JEW5zzeAB6DPI5iImlUMpzqC8cAgEFBnJgwjUcHFFixwGImiwcAA7hsAOTwc0+KCLOCiaXkVD5pqF3MNssVl42hsAARgqgAtO8qgeoNYoEgHU6Xcl3YrlU0VeGRavy0A)

Reply:
Even if I make `StateObject` a unique type, the shape is exactly the same so TS doesn't produce an error. [TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgQQHYE8DKMCGMApgDQpwCiAbkRjAPIBGAVkQMYxmp6Gnk-EcbABZEQ+LnAAqWMEQDmtACLAAzvkYAbIgBNJMuaoLFJqNB2AQMTVh0nmYwKsBhYAsvjD2O0AEpEAM0kAIXxVIgdLaxZ2TnIACTUYaCwANXxNAFc+VAAFKAgvfgBhAA03ABlJHFAwbWpaGFV6IP5jIg8RYAwcuAEiADkIHV7+9KzeqSh8DFUXKKVA7vmrfVkFWgoMDW09cm524qsA4HkyJABfOACCkDgAIgAPI157gG44AHpPuFdDAC44AAWAB0AGYwSCAGxIJDdYhQAL4NhEPrtGyxAA85CkRzoREecVQUgadDgBOIGB0qkoNDoGI4cAAvLTGgyidJ+kJROJyYTaNS0bxuWJ8JjcVZiISAHzMuAzLD6AxEF7EPmUwXK1VEcV4qUwWUshBwKgZbKAhUfNiSimAiX4wlwC76fyqCCaGg6LVuIgEOXKxQYFTqLS6JCy5CoU0TQFjM1EN7ka0OmB2vUUxOoYRJFLjbIAfkBiSMufjcAAPnBMlSlj0dJm4NmS1AsIWhcRsahienCaZpKSOcSuSJRX2pFr2mPXe7Pd7ffhyNKG8jHFZVIC0FBplhMZErOzdTbe-26Qal+QV055iqNxZnK4PGAGyB5xbsA2iKf1yfGgBtAC6H6nnaA4NgA+p+jSxmUlQggO4oDueqBgeEqhzFYwA6ICRhQN08gVnAGCZJomgNp8ABU5H7ORcAAJJUsAbC8DSADuogwKIUC-KIcDao2YRwCIMyKDo1y3NxqJgFAn6WJkNLaiC6C8e0cBqIJa6YUQ0mifcQkYCJ9yqQE-zUdRcAALR0U0JplmpGAQPARAAI6ZBkvwQKp1lSTJEByTZEwkHA0AmV2NGWbR8DCAJCqEUQLHyhYa5wAAFHMIzmYEASxKoACU7l8uwmTECCpmhegGCqRg8xuXxyUsS4wiER5TbJC2eX1SRcDSTAmRQBVAAG1YjCcdb9SVoWfEmUX6bogKMBA7pEDMBFDbWuhkZRZn0TojHMXAbG+pxEnKbwqk0vg1zdDV7Tjag5GTagOhWEQc0Ldoy2VqtI3reQFFUWVUg8bQOy6Cdar2SMNLSd54R0IQTiohAATHXx0bZLdcD3UmxynL18NWBuW74Du-RDCMh4poFCqBSSp5U9g9NYNKSFfJtAM8T0jqQXQNIcYQ+3AJ1TFyaiF0wNMswrBVNwQHcHGomwvXSWSCnUQ9Xw-AAAk05mnPZ0kWXAigwAi1zQNxanGzSyYnPIvWg906mzOLmQ+FxboWzSUmFFpriCctLxYE1GAZURL7TKG5CczAA7fpu24Iaev4AOR-EQyf-izf1mYDqLizMcyrhVw3LEXvNRfA0mqMRxCiY7HFqarE3kPnktl4TCdTAXUuLCNUsU-qNOIVnbN3eVQUxIy4hgGAeEJa1dFKLzHmqGA+AsXW8-QKonzdFQEAANag+EUDOCiqgY1jqBdJoOjK4C-jWlAOiYjheGBQ4fiBJiCrMw2BDyG-DgX0r9xZ4RZuILoPQQpdkrP0To2Yeidi7Cg+0g99goPlAzDBqCBx9lQROXg+CuyhHCHuaIthByYOpjgrsUhpwel0HOAgtDpQYM+jWb69ZYSoGzmVYo0k9oXR6PFfolUXgYBRGbLi8sjYIwGtqPMRB+pYNEv1ZM+oxpmU1mvaYdxFHxm0bo-AdwNEUjVuQGWIBdRD3VAKGkA52RykcZPA0yUMEGJjO2HUuIh6M38VgpmBE4wBQwWYwkbZcQrU4d0MMqAcqxnaDY-sATqaBPPL9UemM4ACKWsQc6sVREqW6BIqRARzayPkPIuA6icbyC0WVHR+A9FO1ttIHifEbanAsdfQRHZfH9jsVSBxp4nEshcZQ6U7iUFdPkIk3geJbbJJJGw+J8z+nFD8YE1J2AMm8KyTRXJQjCk1P6CokpBBJGonKTIniVS6Q1M8dkFRMw1HhJgCo+qHEmrzyiDSVKmkMoBCyhwXKl84BNJaU8ogRjmkmKdvqHplVfbLIoEMwUEzYjONGa4qZHj2hKPWT4zZKTtlkt-sEgl8Y+zvLTOQBJ3jUV7NZv9Me-gep9XOhVZp24grIwyJoFGYD9JwHesjPiENJKEGEBfWFUKqUTDlfCkYmhQAuC0u0hWUVpgcC0qlPKfN4DhGMWqJaIheKZEYGvb59cOnCvwhK4YUqOKX3VskPAuF9KqFSgq803ilGBRVWqhE2F7UMrfvpACiYW4QAAFI4HoAMZKDKG40gAGSIAAd+CN8gAIXGjfs1l2SADq7EjqyMVluRoYNURo1RGpC61dGDhHgEjY69zaBwF0dW1G8ZwWQvhd2ugITshIvEDAEcqhkFDlSOimk9CVQziYesVQPo-QUnsdIdYgYtggx0LQ-McB8iFExMesA4oGGzhXWuiQcBk5VyXToZO0pArJ3HZO-oqhn20MBD0GgUBFzTK7EOmAI6XqclSPS16i0ZgFrgO+0QU6MFSBwLOjdwzpCXuXYYG9c6t1yEDMGPdB7pCEOICnOtGcf2xX-YBjBIGwN2lQ1Bz2Z1GUYOSuOFdKl0OCgVCR40damOpCtEeVMgSnR4a44YdotDUCHukyqWTmCuy-sggBlTan-05RTu8jOY48HIeHDyCQyGyO9GJFhr0175yLjgBmwT8ZhNOjgwhlUyUGO+vA2BhlloeEsrMqWw6GqK1Kx7SpWZeMi78QKY6kYAtvmyNXuwYAJxQb9QAQ0seA65b4HkEiqKqgpB5aA6gABdorPMIurx+d27NjbFDPulTh6z2noKOehdbpGHWZw-OV9D7uvPtfVm79mnlKevy2suA80YMYDg3wseiwEQgFiaxMtGrYY7RFbI-q3MPkCyFvgEW8pg7mRuYwTCIwKqt0LlEMyyRpuonCNoXVehyQPNS8dG7UsaRRRoD8y8SV9ZmT0ooIVp062XyVXcPbmr3vVoe-kmA2juq9VmHAIL8tbmolhx1QVwtwgne+0XJFTEMDJT2yBU8wTaj1C-C0ROjRpQpzThnBlM33pzcuLCAkkBYC-HWDk0QbAD44G2GAVQwgHKYjF54SXDlDR9HF-Lxy-IMP9GQd0LKXFNnkC1xqigkh9dcRwEbjA2vpBm4t74cM5BD39APCSw3fQaaBV8KsuAsZldS5gNGpAIw2CaGaQrNc8BIDMAgP0AAjICYowvRfe+lxr401pqzibDowDVFw-6wmTEYBKKIwCgZUiyZKIBcDtA3NgMRlYM9aTykyCMevkZl4r6dC5MwURtv6HlSMXZy-9AbLwn4AA9fM5ArhEE0ITvvw+gqFFUIFVNcAS5VWIJoIOLErCp0e78KA2RGxaRhSggf7Qh8azgGPiflxo3IkL8X3gHmIAR+jzlRM3xL-j6AA)

Reply:
Even if I change `StateObject` to be a unique non-aliased type, TS still doesn't raise an error: [TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgQQHYE8DKMCGMApgDQpwCiAbkRjAPIBGAVkQMYxmp6Gnk-EcbABZEQ+LnAAqWMEQDmtACLAAzvkYAbIgBNJMuaoLFJqNB2AQMTVh0nmYwKsBhYAsvjD2O0AEpEAM0kAIXxVIgdLaxZ2TnIACTUYaCwANXxNAFc+VAAFKAgvfgBhAA03ABlJHFAwbWpaGFV6IP5jIg8RYAwcuAEiADkIHV7+9KzeqSh8DFUXKKVA7vmrfVkFWgoMDW09cm524qsA4HkyJABfOACCkDgAIgAPI157gG44AHpPuFdDAC44AAWAB0AGYwSCAGxIJDdYhQAL4NhEPrtGyxAA85CkRzoREecVQUgadDgBOIGB0qkoNDoGI4cAAvLTGgyidJ+kJROJyYTaNS0bxuWJ8JjcVZiISAHzMuAzLD6AxEF7EPmUwXK1VEcV4qUwWUshBwKgZbKAhUfNiSimAiX4wlwC76fyqCCaGg6LVuIgEOXKxQYFTqLS6JCy5CoU0TQFjM1EN7ka0OmB2vUUxOoYRJFLjbIAfkBiSMufjcAAPnBMlSlj0dJm4NmS1AsIWhcRsahienCaZpKSOcSuSJRX2pFr2mPXe7Pd7ffhyNKG8jHFZVIC0FBplhMZErOzdTbe-26Qal+QV055iqNxZnK4PGAGyB5xbsA2iKf1yfGgBtAC6H6nnaA4NgA+p+jSxmUlQggO4oDueqBgeEqhzFYwA6ICRhQN08gVnAGCZJomgNp8ABU5H7ORcAAJJUsAbC8DSADuogwKIUC-KIcDao2YRwCIMyKDo1y3NxqJgFAn6WJkNLaiC6C8e0cBqIJa6YUQ0mifcQkYCJ9yqQE-zUdRcAALR0U0JplmpGAQPARAAI6ZBkvwQKp1lSTJEByTZEwkHA0AmV2NGWbR8DCAJCqEUQLHyhYa5wAAFHMIzmYEASxKoACU7l8uwmTECCpmhegGCqRg8xuXxyUsS4wiER5TbJC2eX1SRcDSTAmRQBVAAG1YjCcdb9SVoWfEmUX6bogKMBA7pEDMBFDbWuhkZRZn0TojHMXAbG+pxEnKbwqk0vg1zdDV7Tjag5GTagOhWEQc0Ldoy2VqtI3reQFFUWVUg8bQOy6Cdar2SMNLSd54R0IQTiohAATHXx0bZLdcD3UmxynL18NWBuW74Du-RDCMh4poFCqBSSp5U9g9NYNKSFfJtAM8T0jqQXQNIcYQ+3AJ1TFyaiF0wNMswrBVNwQHcHGomwvXSWSCnUQ9Xw-AAAk05mnPZ0kWXAigwAi1zQNxanGzSyYnPIvWg906mzOLmQ+FxboWzSUmFFpriCctLxYE1GAZURL7TKG5CczAA7fpu24Iaev4AOR-EQyf-izf1mYDqLizMcyrhVw3LEXvNRfA0mqMRxCiY7HFqarE3kPnktl4TCdTAXUuLCNUsU-qNOIVnbN3eVQUxIy4hgGAeEJa1dFKLzHmqGA+AsXW8-QKonzdFQEAANag+EUDOCiqgY1jqBdJoOjK4C-jWlAOiYjheGBQ4fiBJiCrMw2BDyG-DgX0r9xZ4RZuILoPQQpdkrP0To2Yeidi7Cg+0g99goPlAzDBqCBx9lQROXg+CuyhHCHuaIthByYOpjgrsUhpwel0HOAgtDpQYM+jWb69ZYSoGzmVfwPU+rnQqvgImQckbyk6vLZSuF9JwHesjPiENJKEGEBfMyms17TDuNqPMRANFaPwHcEYmhQAuC0tIHiQlpgcC0qlPKfN4DhEMWqJaIheKZEYGvDilUUZgLkUo4YKiOKX3VskPAsjAGpXaHo2MMT4yBRMWYhE2F-HyByqkyJAFEwtwgAAKRwPQAYyUMmezgAAMkQAA78b99IAQuDk3ho9MZwAAOrsSOtIxWW5Ghg1RGjVEakLrV0YOEeAEjpHyARhVLRvTUbxkvnATRoijFwFmXQOMEw1bkHEDAEcqhkFDlSOqAUNJ6EqhnEw9YqgfR+gpKc6Q6xAxbBBjoWh+Y4D5EKJiL5YBxQMNnNc25Eg4DJyrpcnQydpSBWTrs-Z-RVBQtoYCHoNAoCLmShg9ZMBNnmk5KkcgpT5qLRmI0uAcLRAHIwVIHAxz7lUjOQCq5hhgUnIZY8uQgZgyvPedIQhxAU4DIzsi2KaKMVYtEY0XFL18WEsBA3Gkal+iHJSuOa5Kl6WCgVLy40Ay7S0qtEeVMWCg5XE1Wc-l+jMFdg+Wqww7RaGoBRZBdF1rnVopyinZM+oM5jjwdS4cPIJDUstVOC5jCvRAvnIuCpiB-J4ppcchpPDyWEH2clbF0q4m8D0aUy0Ka+Fj3aYdCxXSlZzJUjbXG0wi78XOn0pqIwBY+OkavdgwATig36gAsaBiVly3wPIbZWYwhSEHZilBAC7RMsjSy+cbLNRPM2NsUMbzrUfN+T8gofzzlugjcwkFYLw2eihTC6pSK3UyLwnKuAxL3oYDJYWlpiwEQgG6CqfaHSLGwx2nI6R-VuYwH6gLIW+ARbymDuZAI0BGCYRGBVVuhcohmWSLe1E4RtC2L0OSOkRljqIaljSKKNAmrzyiDSfWZk9KKD8adAZl8+3aJw703OzGySoeIEYDR3VeqzDaV+ri0jAMgc0P7cDYsJZIasMO-2GBkqAZAqeAiNRwD1C-C0ROjRpQpzThnIlb0loPsuLCAkkBYC-HWHAYoog2AHxwNsMAqhhAOUxPZzwTmHKGj6A5jzjl+TsuVeQboWUuLFEkMFixFBwsYBC30aLsWpDxYsb4cM5APn9APGFyggUcA00Cr4NhTrvPueczAHJSARhsE0JKp2Rg1kQGYBAfoABGQE1n2B2Z86V1z7RMTGmtNWY1YdGAWIuH-WEyY6vIhRGAHFKkWTJRALgdoG5sD9AIsNrSeUmQRiC8jRby3TrdBeBgFEEj+h5UjF2Jb-QGy8J+AAPXzOQM1mhwiIAwd8IKhRVCBQVXAEuVViCaCDixKwqc0O-CgNkRsWkrXXcO8QO7Gs4BPZe5cHJ02iCzf6BmhrTX2jNZyomL7aOgA)

Reply:
Ah, you are right here. This class has no private members and thus the structural type system allows this POJO argument to be passed there. Within the function, the TS assumes that it can narrow down the union using this `instaceof`. This is not sound but I guess that it works for practical reasons.

The POJO thing still behaves a little bit differently before we pass it to some other context:
[TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgUQG4FMB2MDyARgFZYDGMANCnAMowCGMW1q9TWtZAFliA6zgAVAJ5gsAc1wARYAGcGhADZYAJoNHi5jZoICCFYBBxFSFfYYzAYIgLIMwFmNABKWAGaCAQgzlYDMEYmJORUNAAS8s5QIgBqDEoAriw0AApQEI40tADCABq2ADKCtKBgKpi4MHL4ntk6WPY8wDgpbA0AchCqbXQN8Um9QlAMOHLWQdIeLRPGGmKSuGg4iirqSAC+cO4ZIHAARAAe2hz7ANxwAPSXcDZaAFxwACwAdADMby8AbEhILcxQdwMMhYPocUyhAA8NCEOWMzEOYVQQkqeDgWERuFUcnQ2DwEIocAAvLiqgSkcJ2MwuLx+OjMThsWDqTw+AxIbD4RiYAA+YlwUYiDSaLAnZj05iMnEisVYDlwvDcvkkhBwDAJZKPQUXMhcxGPTmKxFwDYaNxyCBKbCqGW2LCMfkiqQ4WQKZRqJB85CodWDR5UrADZJnGi6o0wA0KhEwEOobhRaBxDVYAD8j0i2kTQdBAB84IlGdNWqpY3B45mYmnmXKaMio9zBMjURTkQGaWzG8IZQ1O0JzZbrbb7QwaDzS8DAsY5I89FARiJIQEguT5XqqMJmzyxzQJ8ArIFRTPLNY7A5SyBh1qcCJS1g8dUDc2ANoAXVv98f99LAH071V-fkRQvM2HKbj+fhyOMxjAKojzaFALQSHAeY4IkShKKWlwAFRYbWWFwAAkoywBkBwOIAO68DAvBQLcvBwLKZa+HAPCjFIqjbLsdGgmAUB3kYiQ4rKLxwHoDENHA8gsVOMFYHxHH7KxODsfsknuPceF4XAAC0hHVGqyaSTiOAQPAWAAI6JAktwQJJ+m8fxECCQZgyUHA0AaagcD4bpBHwNwzGCnArTkQKhhTnAAAU4w9NpHjuKEcgAJQ2fS5CJMwLyaV5+F6Dgkk4BM1mMZF5HWNwwW2eW0QiClZXoXAfEwIkUD5QABgWPTuC0ahtVlOWXKGAXKWojyEBAlpYKMyH5oW3XFphOFaURqgkWRcCUfaNHceJHBGQK2wtMVDT9agWGDagqjGFgY0TSo015p1RZqItuE5cI9G4Ksai7eKJk9DifEOX4eBMHuoIQO4O2Mb6ySnd5F3STg3USC1YPGDOc4MAuAZdD0q7hm5gpuSi95E9e5MiFur1aUI9GtMaf54Di1FMBtwANaRgmggwtwjGMsz5TsEB7NRoJkC1fFosJeGI9ccAAALVNpwASCZfE6XAUgwAC2zQHRUnaziYYoy1P0tEj8GJBQ+sWgbOK8Zkck2Cx00nCIlU4HFqEXiM7o0AzMDNtOolYwupNVE++x3Fg+wvtuqDYW9Z0faCMD8+Mk75V1MxZyzAXwHxchocwHEW9RUkywNNDp6MmdBCHs7zhyGeC1M82CwT0Yk5uCdXEt715e5ISEvwYBgIhYXRIR0gs7ZchgAw5HFlP0ByJcLQYBAADWP1+FAVggnI8PnUNHOqFLjxuLqUCqJC8GIW5ASuB4kKCtTNcMBIIe0Pa9-p4hPu-BmitE8l5PMAYmjxlaNCLycDhD1kRJ2OBxNazwIjngZBXkhDdg4Fg1APg-BLmMOSfBAoKZoLgX2UUA41BDkYJQuAPI0GPTmj1EsvxE4DxTm4ZqrUcTTQYGHdyUMEhKGhgA5ScB7pQ0Yv9HiTBuDHy0grReIw9iymzCotRDA9g9CUKAawclU4sQCiMCgcloopVZvAPwOjxRTR4AxRIhBF7UQKhIhCUi5HdAUdRE+iNnD0C8d-aK-Rkz+nCa5OA+jDEAjgpIiQSUEkhNfCGGuEAABStB8AdEisk+2cAABkiBbhfxDg-ZSr4TTpK4cnbycAADqVFtpixYpLKov1QSw1BFJXmxdCB+HgJDHaEhwb5TUZ0mGyYT6Kx0XsSZeAAxaOrqgfgMBWRyFgdg2gsQJRYmlP2K0dCFhyDtA6bkBzhALGdMsb6qhGEpjgOkTIkIXlgA5EcwcpzzkCAOEXWhqh9g8jcvsdZmyAxyGBYwx4rRsBQFHJFNBiyYDLIiZSWINACnjUmqMWpcBwW8C2WgoQuz9lSmEF8k5WhfnkqZE6GQ8h7mPK7KchoUcelxxhcFP8CLUA8iRXAlFaK-QYqxY8CuOIpIBm2agSKOC2V7UuRSwULLVQ9INLsnUa4rwey2Mq+luDmCMNQE8hVWgGgmrgLC3l3K4VySSlHMM0Y469mbL2NsrJ+C9iNb0ZEVKbQ-OHKOYppSNUYpqZwglTBNmRWFVEzU1ZswFO1FGpOWlmlbWMW0iWc4pkSRNqrNGWcmICK6ZVHo7N3FtIXuQYA3UfptUYBIPq2ihG6LKRIWWNAApyCEF-QVXlm0GgDfQ3mBrpQ3KWCsd0Dz4FwCee8t5GQPnUItMcwNNLhygoBRu4FoLm1QuYfOlJiFxVwBxfdHA+L03vSmACEAPUKItOMSDVaUi2ltSZjANq7NOYMG5gdEyXt3DQEIDBHo+Va4CyzlpZwF7QR+BUBY9Q6I8RqR2tB+uEUArYEqlPBulU+UpyUlICRe0eknzbeotDnS6agm-alZg2gVFNRamMJpL7aJtMY-VcRXM-AHSw4LbtqBSI4Eit+z8nSIFlAqPeGo7hQL3h5FHGOcdsV3SmtezYvwMSQFgLcBYcAci8DINvWgKwwByG4KZSElmHA2dMsqOgVmnNmQZEyGVNAWgJVojkQQvnjFoEC8jYxtBQt+eEJF4xLhPQ0CeQGFcAX0BuQi9FuALhj3Wtc452zMZfg9DIEoIR4spzwEgMQCAAZHimfIBZtz+X7MNEhKqXUBYIzBUSCAQgxiNhbl+PWqKlXqsSRaCcHAIIRkBhSt6OAI2AyYRuAAPRTJsdEShBNzYWw0JbcBVu6aQEVkrGswzaC6URCbIJIkcFa9JDrsLuu9dov19JQ2wkcEu4wSbvSxjfam1DGbiAaCyi+6MEEe2Dv6s26COboO-vg6wJDtbGwkBAA)

It's also more honest to the reader and serves as a documentation~. So I think that I would still like to use it.

To avoid copy-pasting everything from XState you could create it like this:
```ts
type StateObject<T extends AnyState> = { [K in keyof T]: T[K] }

// if we want to keep the same generic arguments then we can do this:
type IdentityRemap<T> = { [K in keyof T]: T[K] };
type StateObject<
  TContext,
  TEvent extends EventObject = EventObject,
  TStateSchema extends StateSchema<TContext> = any,
  TTypestate extends Typestate<TContext> = { value: any; context: TContext },
  TResolvedTypesMeta = TypegenDisabled
> = IdentityRemap<
  State<TContext, TEvent, TStateSchema, TTypestate, TResolvedTypesMeta>
>;
```

The second variant doesn't display well in the IDE though as `StateObject` doesn't get printed and instead we see `IdentityRemap<State<...>>`. 

To fix that we could use this:
```ts
type StateObject<
  TContext,
  TEvent extends EventObject = EventObject,
  TStateSchema extends StateSchema<TContext> = any,
  TTypestate extends Typestate<TContext> = { value: any; context: TContext },
  TResolvedTypesMeta = TypegenDisabled
> = Pick<
  State<TContext, TEvent, TStateSchema, TTypestate, TResolvedTypesMeta>,
  keyof AnyState
>;
```